odoo_server_maintenance/controllers/main.py

```diff
@@ -0,0 +1,26 @@
+# -*- coding: utf-8 -*-
+from openerp import http
+from openerp.http import request
+from openerp.tools.translate import _
+import openerp.addons.web.controllers.main as main
+
+
+class Home(main.Home):
+
+    @http.route('/web/login', type='http', auth="none")
+    def web_login(self, redirect=None, **kw):
+        values = request.params.copy()
+        if request.httprequest.method == 'POST':
+            uid = request.session.authenticate(request.session.db,
+                                               request.params['login'],
+                                               request.params['password'])
+            if uid is not False:
+                maintenance_mode = \
+                    request.env.user.browse(uid).maintenance_mode
+                if maintenance_mode:
+                    values['error'] = _("This user is in maintenance mode")
+                    if request.env.ref('web.login', False):
+                        return request.render('web.login', values)
+        return super(Home, self).web_login(redirect, **kw)
```

Comment:
AFAIK "routes" are lazy evaluated, so the redirect will not done until all inherited methods are run.
This means you could have the "super" call first, and then check for the maintenance mode. 
With this, no need to perform the authentication twice.